pagerduty/resource_pagerduty_custom_field.go

```diff
@@ -0,0 +1,325 @@
+	}
+	return nil
+}
+
+func resourcePagerDutyCustomFieldDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, err := meta.(*Config).Client()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	_, err = client.CustomFields.DeleteContext(ctx, d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	return nil
+}
+
+func resourcePagerDutyCustomFieldUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, err := meta.(*Config).Client()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	field, err := buildFieldStruct(d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	log.Printf("[INFO] Updating PagerDuty field %s", d.Id())
+
+	updatedField, _, err := client.CustomFields.UpdateContext(ctx, d.Id(), field)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	err = flattenCustomField(d, updatedField)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	return nil
+}
+
+func resourcePagerDutyCustomFieldRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	log.Printf("[INFO] Reading PagerDuty field %s", d.Id())
+	err := fetchField(ctx, d, meta, handleNotFoundError)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	return nil
+}
+
+func fetchField(ctx context.Context, d *schema.ResourceData, meta interface{}, errorCallback func(error, *schema.ResourceData) error) error {
+	client, err := meta.(*Config).Client()
+	if err != nil {
+		return err
+	}
+
+	return resource.RetryContext(ctx, 2*time.Minute, func() *resource.RetryError {
+		field, _, err := client.CustomFields.GetContext(ctx, d.Id(), nil)
+		if err != nil {
+			log.Printf("[WARN] Field read error")
+			errResp := errorCallback(err, d)
+			if errResp != nil {
+				time.Sleep(2 * time.Second)
+				return resource.RetryableError(errResp)
+			}
+
+			return nil
+		}
+
+		if err := flattenCustomField(d, field); err != nil {
+			return resource.NonRetryableError(err)
+		}
+		return nil
+
+	})
+}
+
+func flattenCustomField(d *schema.ResourceData, field *pagerduty.CustomField) error {
+	d.SetId(field.ID)
+	d.Set("name", field.Name)
+	if field.Description != nil {
+		d.Set("description", *(field.Description))
+	}
+	d.Set("display_name", field.DisplayName)
+	d.Set("datatype", field.DataType.String())
+	d.Set("multi_value", field.MultiValue)
+	d.Set("fixed_options", field.FixedOptions)
+	return nil
+}
+
+func buildFieldStruct(d *schema.ResourceData) (*pagerduty.CustomField, error) {
+	field := pagerduty.CustomField{
+		Name:         d.Get("name").(string),
+		DisplayName:  d.Get("display_name").(string),
+		DataType:     pagerduty.CustomFieldDataTypeFromString(d.Get("datatype").(string)),
+		MultiValue:   d.Get("multi_value").(bool),
+		FixedOptions: d.Get("fixed_options").(bool),
+	}
+	if desc, ok := d.GetOk("description"); ok {
+		str := desc.(string)
+		field.Description = &str
+	}
+	return &field, nil
+}
+
+func validateFieldDataType() schema.SchemaValidateDiagFunc {
```

Comment:
Could you please move the utility functions that are being re-used for this resource, on custom field options and schemas to a new file with that purpose? in a fashion like `event_orchestration_path_util.go`

Reply:
Great idea! Hadn't seen this pattern applied in this provider code so didn't want to invent anything new. Moved the functions used in more than one resource to `custom_field_util.go` and the unit tests for those (although most of the utility code is tested via acceptance tests) in `custom_field_util_test.go`.